File: OneXray-temp/OneXray-main/android/app/src/main/kotlin/net/yuandev/onexray/pigeon/HostApi.kt
package net.yuandev.onexray.pigeon

import android.Manifest
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.net.VpnService
import android.os.Build
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.FragmentActivity
import com.elvishew.xlog.XLog
import com.hjq.permissions.Permission
import com.hjq.permissions.XXPermissions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import libXray.LibXray
import net.yuandev.onexray.vpn.VpnController
import net.yuandev.onexray.vpn.VpnStatusConnection
import java.io.ByteArrayOutputStream

class AppHostApi(
    private val context: Context,
) : BridgeHostApi {
    private val vpnStatus = VpnStatusConnection(context) { status ->
        scope.launch { flutterApi?.vpnStatusChanged(status) }
    }
    private val activity = context as FragmentActivity
    private val prepareResult =
        activity.registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            val callback = permissionCallback
            permissionCallback = null
            if (it.resultCode == RESULT_OK) {
                callback?.invoke(queryPermissionNow())
            } else {
                callback?.invoke(androidPermissionDenied())
                onVpnStatusChanged(false)
            }
        }

    fun onVpnStatusChanged(running: Boolean, error: String? = null) {
        XLog.d("AppHostApi: onVpnStatusChanged running=$running")
        scope.launch(Dispatchers.Main.immediate) {
            try { vpnStatus.changed(running, error) }
            catch (failure: Exception) {
                XLog.e("VPN status notification failed", failure)
            }
        }
    }

    private var flutterApi: AppFlutterApi? = null

    fun onInit(api: AppFlutterApi) {
        XLog.init()
        flutterApi = api
        scope.launch {
            try { api.vpnStatusChanged(vpnStatus.read()) }
            catch (error: Exception) { XLog.e("Initial VPN status failed", error) }
        }
    }

    fun onDestroy() {
        vpnStatus.close()
        scope.cancel()
    }

    private var permissionCallback: ((PlatformPermissionResult) -> Unit)? = null

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val invokeMutex = Mutex()

    override fun getTunFilesDir(callback: (Result<String>) -> Unit) {
        val dirPath = context.filesDir.path
        callback(Result.success(dirPath))
    }

    override fun readVpnStatus(callback: (Result<NativeVpnCommandResult>) -> Unit) {
        scope.launch {
            try {
                val status = vpnStatus.read()
                callback(Result.success(NativeVpnCommandResult(
                    state = NativeVpnCommandState.SUCCESS,
                    permission = queryPermissionNow(),
                    status = status,
                    message = if (status == VpnStatus.DISCONNECTED) VpnController.lastError else null
                )))
            } catch (error: Exception) {
                VpnController.lastError = error.message ?: error.toString()
                callback(Result.success(commandFailed(queryPermissionNow())))
            }
        }
    }

    override fun startVpn(callback: (Result<NativeVpnCommandResult>) -> Unit) {
        XLog.d("AppHostApi: startVpn called")
        scope.launch {
            val permission = queryPermissionNow()
            if (permission.state != PlatformPermissionState.GRANTED) {
                callback(Result.success(waitingForPermission(permission)))
                return@launch
            }
            try {
                val status = vpnStatus.command(VpnStatus.CONNECTED) { VpnController.startVpn(context) }
                callback(Result.success(commandSuccess(permission, status)))
            } catch (error: Exception) {
                VpnController.lastError = error.message ?: error.toString()
                callback(Result.success(commandFailed(permission)))
            }
        }
    }

    override fun stopVpn(callback: (Result<NativeVpnCommandResult>) -> Unit) {
        XLog.d("AppHostApi: stopVpn called")
        scope.launch {
            try {
                val status = vpnStatus.command(VpnStatus.DISCONNECTED) { VpnController.stopVpn(context) }
                callback(Result.success(commandSuccess(queryPermissionNow(), status)))
            } catch (error: Exception) {
                VpnController.lastError = error.message ?: error.toString()
                callback(Result.success(commandFailed(queryPermissionNow())))
            }
        }
    }

    override fun invoke(requestJson: String, callback: (Result<String>) -> Unit) {
        scope.launch {
            // Temporary cores share process-global Xray state. The VPN runs in :native.
            invokeMutex.withLock {
                callback(runCatching { LibXray.invoke(requestJson) })
            }
        }
    }

    override fun queryPlatformPermission(callback: (Result<PlatformPermissionResult>) -> Unit) {
        scope.launch {
            callback(Result.success(queryPermissionNow()))
        }
    }

    override fun requestPlatformPermission(callback: (Result<PlatformPermissionResult>) -> Unit) {
        scope.launch {
            val permission = queryPermissionNow()
            if (permission.state == PlatformPermissionState.GRANTED) {
                callback(Result.success(permission))
                return@launch
            }
            val prepare = VpnService.prepare(context)
            if (prepare == null) {
                callback(Result.success(queryPermissionNow()))
                return@launch
            }
            if (permissionCallback != null) {
                callback(
                    Result.success(
                        PlatformPermissionResult(
                            PlatformPermissionKind.ANDROID_VPN,
                            PlatformPermissionState.AWAITING_USER_APPROVAL,
                            "Android VPN permission is already pending.",
                        )
                    )
                )
                return@launch
            }
            permissionCallback = { result ->
                callback(Result.success(result))
            }
            activity.runOnUiThread {
                prepareResult.launch(prepare)
            }
        }
    }


    override fun getInstalledApps(callback: (Result<List<AndroidAppInfo>>) -> Unit) {
        scope.launch {
            checkInstalledAppPermission {
                if (it) {
                    val packageManager = context.packageManager
                    val installedApps = packageManager.getInstalledApplications(0)
                    val apps = mutableListOf<AndroidAppInfo>()
                    for (info in installedApps) {
                        val appInfo =
                            AndroidAppInfo(
                                packageManager.getApplicationLabel(info).toString(),
                                info.packageName,
                            )
                        apps.add(appInfo)
                    }
                    callback(Result.success(apps))
                } else {
                    callback(Result.success(listOf()))
                }
            }
        }
    }

    override fun getAppIcon(packageName: String, callback: (Result<ByteArray?>) -> Unit) {
        scope.launch {
            callback(Result.success(loadAppIconBytes(packageName)))
        }
    }

    private fun checkInstalledAppPermission(callback: (Boolean) -> Unit) {
        // android 11, level 30
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val permissions = mutableListOf<String>()
            permissions.add(Manifest.permission.QUERY_ALL_PACKAGES)
            permissions.add(Permission.GET_INSTALLED_APPS)
            XXPermissions.with(context)
                .permission(permissions)
                .request { _, allGranted ->
                    callback(allGranted)
                }
        } else {
            callback(true)
        }
    }

    private fun loadAppIconBytes(packageName: String): ByteArray? {
        return try {
            val drawable = context.packageManager.getApplicationIcon(packageName)
            val bitmap = Bitmap.createBitmap(ICON_SIZE_PX, ICON_SIZE_PX, Bitmap.Config.ARGB_8888)
            try {
                drawable.setBounds(0, 0, ICON_SIZE_PX, ICON_SIZE_PX)
                drawable.draw(Canvas(bitmap))
                val stream = ByteArrayOutputStream()
                if (bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)) {
                    stream.toByteArray()
                } else {
                    null
                }
            } finally {
                bitmap.recycle()
            }
        } catch (e: Exception) {
            XLog.d("AppHostApi: getAppIcon failed for $packageName: $e")
            null
        }
    }

    // macOS
    override fun appleVpnCapabilities(callback: (Result<AppleVpnCapabilities>) -> Unit) {
        callback(Result.success(AppleVpnCapabilities(false, false)))
    }

    override fun useSystemExtension(callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(false))
    }

    override fun queryLaunchAtLogin(callback: (Result<NativeLaunchAtLoginResult>) -> Unit) {
        callback(
            Result.success(
                NativeLaunchAtLoginResult(
                    NativeLaunchAtLoginState.UNAVAILABLE,
                    null,
                )
            )
        )
    }

    override fun setLaunchAtLogin(
        enabled: Boolean,
        callback: (Result<NativeLaunchAtLoginResult>) -> Unit,
    ) {
        callback(
            Result.success(
                NativeLaunchAtLoginResult(
                    NativeLaunchAtLoginState.UNAVAILABLE,
                    null,
                )
            )
        )
    }

    override fun openLaunchAtLoginSettings(callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(false))
    }

    //ios
    override fun setAppIcon(appIcon: String, callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(true))
    }

    override fun getCurrentAppIcon(callback: (Result<String>) -> Unit) {
        callback(Result.success(""))
    }

    private fun queryPermissionNow(): PlatformPermissionResult {
        val prepare = VpnService.prepare(context)
        if (prepare != null) {
            return PlatformPermissionResult(
                PlatformPermissionKind.ANDROID_VPN,
                PlatformPermissionState.NOT_DETERMINED,
                null,
            )
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.CINNAMON_BUN &&
            context.checkSelfPermission(Manifest.permission.ACCESS_LOCAL_NETWORK) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            return PlatformPermissionResult(
                PlatformPermissionKind.ANDROID_LOCAL_NETWORK,
                PlatformPermissionState.NOT_DETERMINED,
                null,
            )
        }
        return androidPermissionGranted()
    }

    private fun commandFailed(permission: PlatformPermissionResult): NativeVpnCommandResult =
        NativeVpnCommandResult(
            state = NativeVpnCommandState.FAILED,
            permission = permission,
            message = VpnController.lastError,
        )

    private fun androidPermissionGranted() = PlatformPermissionResult(
        PlatformPermissionKind.ANDROID_VPN,
        PlatformPermissionState.GRANTED,
        null,
    )

    private fun androidPermissionDenied() = PlatformPermissionResult(
        PlatformPermissionKind.ANDROID_VPN,
        PlatformPermissionState.DENIED,
        null,
    )

    private fun commandSuccess(permission: PlatformPermissionResult, status: VpnStatus) = NativeVpnCommandResult(
        state = NativeVpnCommandState.SUCCESS,
        status = status,
        permission = permission,
    )

    private fun waitingForPermission(permission: PlatformPermissionResult) = NativeVpnCommandResult(
        state = NativeVpnCommandState.WAITING_FOR_PLATFORM_PERMISSION,
        permission = permission,
    )

    private companion object {
        // The per-app list renders icons in a 31dp slot; 96px covers 3x density.
        const val ICON_SIZE_PX = 96
    }
}

File: swift/App/pigeon/HostApi.swift
import Foundation
#if os(iOS)
import Flutter
#elseif os(macOS)
import AppKit
import FlutterMacOS
#else
#error("Unsupported platform.")
#endif

@MainActor
final class AppHostApi: @preconcurrency BridgeHostApi {
    private let flutterApi: AppFlutterApi
    init(flutterApi: AppFlutterApi) {
        self.flutterApi = flutterApi
    }
    
    func getTunFilesDir(completion: @escaping (Result<String, any Error>) -> Void) {
        if let groupUrl = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId()) {
            let path = groupUrl.adaptedPath()
            completion(.success(path))
        } else {
            completion(.success(""))
        }
    }

    func readVpnStatus(completion: @escaping (Result<NativeVpnCommandResult, any Error>) -> Void) {
        Task {
            let permission = await VPNManager.shared.queryPlatformPermission()
            do {
                let status = try await flutterApi.readVpnStatus()
                if permission.state == .failed {
                    completion(.success(NativeVpnCommandResult(
                        state: .failed,
                        permission: permission,
                        message: permission.message
                    )))
                } else {
                    completion(.success(NativeVpnCommandResult(
                        state: .success,
                        status: status,
                        permission: permission,
                        message: status == .disconnected ? VPNManager.shared.lastCommandError : nil
                    )))
                }
            } catch {
                completion(.success(NativeVpnCommandResult(
                    state: .failed,
                    permission: permission,
                    message: error.localizedDescription
                )))
            }
        }
    }
    
    func startVpn(completion: @escaping (Result<NativeVpnCommandResult, any Error>) -> Void) {
        Task {
            let installed = await VPNManager.shared.startVpn()
            let permission = await VPNManager.shared.queryPlatformPermission()
            flutterApi.refreshVpn(result: installed)
            completion(.success(await commandResult(installed, permission: permission)))
        }
    }

    func stopVpn(completion: @escaping (Result<NativeVpnCommandResult, any Error>) -> Void) {
        Task {
            let installed = await VPNManager.shared.stopVpn()
            let permission = await VPNManager.shared.queryPlatformPermission()
            flutterApi.refreshVpn(result: installed)
            completion(.success(await commandResult(installed, permission: permission)))
        }
    }
    
    func invoke(requestJson: String, completion: @escaping (Result<String, any Error>) -> Void) {
        Task {
            do {
                completion(.success(try await LibXrayInvoker.invoke(requestJson)))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func queryPlatformPermission(completion: @escaping (Result<PlatformPermissionResult, any Error>) -> Void) {
        Task {
            completion(.success(await VPNManager.shared.queryPlatformPermission()))
        }
    }

    func requestPlatformPermission(completion: @escaping (Result<PlatformPermissionResult, any Error>) -> Void) {
        Task {
            completion(.success(await VPNManager.shared.requestPlatformPermission()))
        }
    }
    
    /// android
    func getInstalledApps(completion: @escaping (Result<[AndroidAppInfo], any Error>) -> Void) {
        completion(.success([]))
    }

    func getAppIcon(packageName: String, completion: @escaping (Result<FlutterStandardTypedData?, any Error>) -> Void) {
        completion(.success(nil))
    }

    /// macOS
    func useSystemExtension(completion: @escaping (Result<Bool, any Error>) -> Void) {
        completion(.success(Constants.useSystemExtension))
    }

    func appleVpnCapabilities(completion: @escaping (Result<AppleVpnCapabilities, any Error>) -> Void) {
        let serviceExclusions: Bool
        let deviceCommunication: Bool
        if #available(iOS 16.4, macOS 13.3, *) {
            serviceExclusions = true
        } else {
            serviceExclusions = false
        }
        if #available(iOS 17.4, macOS 14.4, *) {
            deviceCommunication = true
        } else {
            deviceCommunication = false
        }
        completion(.success(AppleVpnCapabilities(
            serviceExclusions: serviceExclusions,
            deviceCommunication: deviceCommunication
        )))
    }

    func queryLaunchAtLogin(
        completion: @escaping (Result<NativeLaunchAtLoginResult, any Error>) -> Void
    ) {
#if os(macOS)
        completion(.success(LaunchAtLoginService.query()))
#else
        completion(.success(NativeLaunchAtLoginResult(
            state: .unavailable,
            message: nil
        )))
#endif
    }

    func setLaunchAtLogin(
        enabled: Bool,
        completion: @escaping (Result<NativeLaunchAtLoginResult, any Error>) -> Void
    ) {
#if os(macOS)
        completion(.success(LaunchAtLoginService.setEnabled(enabled)))
#else
        completion(.success(NativeLaunchAtLoginResult(
            state: .unavailable,
            message: nil
        )))
#endif
    }

    func openLaunchAtLoginSettings(
        completion: @escaping (Result<Bool, any Error>) -> Void
    ) {
#if os(macOS)
        completion(.success(LaunchAtLoginService.openSettings()))
#else
        completion(.success(false))
#endif
    }

    /// Apple app icon
    func setAppIcon(appIcon: String, completion: @escaping (Result<Bool, any Error>) -> Void) {
#if os(iOS)
        var iconName: String? = appIcon
        if appIcon.isEmpty {
            iconName = nil
        }
        if UIApplication.shared.alternateIconName == iconName {
            completion(.success(true))
            return
        }
        UIApplication.shared.setAlternateIconName(iconName) { error in
            if let error = error {
                YGLog(error.localizedDescription)
                completion(.failure(error))
            } else {
                completion(.success(true))
            }
        }
#elseif os(macOS)
        completion(.success(DockIconService.setIcon(appIcon)))
#endif
    }

    func getCurrentAppIcon(completion: @escaping (Result<String, any Error>) -> Void) {
#if os(iOS)
        var appIcon = ""
        if let iconName = UIApplication.shared.alternateIconName {
            appIcon = iconName
        }
        completion(.success(appIcon))
#elseif os(macOS)
        completion(.success(DockIconService.currentIconName))
#endif
    }

    private func commandResult(
        _ result: RefreshVpnResult,
        permission: PlatformPermissionResult
    ) async -> NativeVpnCommandResult {
        switch result {
        case .installed:
            do {
                return NativeVpnCommandResult(
                    state: .success,
                    status: try await flutterApi.readVpnStatus(),
                    permission: permission,
                    message: nil
                )
            } catch {
                return NativeVpnCommandResult(state: .failed, permission: permission, message: error.localizedDescription)
            }
        case .waitForApproval:
            return NativeVpnCommandResult(
                state: .waitingForPlatformPermission,
                permission: permission,
                message: nil
            )
        case .notInstalled:
            if permission.state == .awaitingUserApproval || permission.state == .notDetermined {
                return NativeVpnCommandResult(
                    state: .waitingForPlatformPermission,
                    permission: permission,
                    message: nil
                )
            }
            return NativeVpnCommandResult(
                state: .failed,
                permission: permission,
                message: VPNManager.shared.lastCommandError ?? permission.message
            )
        }
    }

    private func commandSuccess(permission: PlatformPermissionResult) -> NativeVpnCommandResult {
        NativeVpnCommandResult(
            state: .success,
            permission: permission,
            message: nil
        )
    }
}
